compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/ExpressionCodegen.kt

```diff
@@ -647,55 +641,53 @@ class ExpressionCodegen(
-        if (!genericOrAnyOverride) return
-
-        // Result parameter of SAM-wrapper to Java SAM is already unboxed in visitGetValue, do not unbox it anymore
-        if (irFunction.parentAsClass.superTypes.any { it.getClass()?.isFromJava() == true }) return
-
-        // Do not unbox Results in suspend lambda `invoke` methods. These just forward to `invokeSuspend`,
-        // where the arguments are unboxed.
-        if (
-            irFunction.parentAsClass.origin == JvmLoweredDeclarationOrigin.SUSPEND_LAMBDA &&
-            irFunction.name == OperatorNameConventions.INVOKE
-        ) return
-
-        StackValue.unboxInlineClass(OBJECT_TYPE, arg.type.erasedUpperBound.defaultType, mv, typeMapper)
-    }
-
-    private fun IrClass.isSamAdapter(): Boolean = this.superTypes.any { it.getClass()?.isFun == true }
-
-    private fun onlyResultInlineClassParameters(): Boolean = irFunction.valueParameters.all { !it.type.requiresMangling }
+    internal fun genOrGetLocal(expression: IrExpression, type: Type, parameterType: IrType, data: BlockInfo): StackValue =
+        if (expression is IrGetValue)
+            StackValue.local(
+                findLocalIndex(expression.symbol), frameMap.typeOf(expression.symbol),
+                expression.symbol.owner.realType.toIrBasedKotlinType()
+            )
+        else
+            gen(expression, type, parameterType, data)
 
-    private fun hasBridge(): Boolean = irFunction.parentAsClass.declarations.any { function ->
-        function is IrFunction && function != irFunction &&
-                context.methodSignatureMapper.mapSignatureSkipGeneric(function).let {
-                    it.asmMethod.name == signature.asmMethod.name && it.valueParameters == signature.valueParameters
-                }
+    // We do not mangle functions if Result is the only parameter of the function. This means that if a function
+    // taking `Result` as a parameter overrides a function taking `Any?`, there is no bridge unless needed for
+    // some other reason, and thus `Result` is actually `Any?`. TODO: do this stuff at IR level?
+    val IrValueDeclaration.realType: IrType
+        get() = parent.let { parent ->
+            val isBoxedResult = this is IrValueParameter && parent is IrSimpleFunction &&
+                    parent.dispatchReceiverParameter != this &&
+                    (parent.parent as? IrClass)?.fqNameWhenAvailable != StandardNames.RESULT_FQ_NAME &&
+                    parent.resultIsActuallyAny(index) == true
+            return if (isBoxedResult) context.irBuiltIns.anyNType else type
+        }
+
+    // Argument: null for return value, -1 for extension receiver, >= 0 for value parameter.
+    //           (It does not make sense to check the dispatch receiver.)
+    // Return: null if this is not a `Result<T>` type at all, false if this is an unboxed `Result<T>`,
+    //         true if this is a `Result<T>` overriding `Any?` and so it is boxed.
```

Comment:
This is a somewhat stupid use of a ternary value, but this entire thing is stupid so anything prettier is too much effort.